uast/diff/diff.go

```diff
@@ -0,0 +1,287 @@
+package diff
+
+import (
+	"fmt"
+	"github.com/heetch/lapjv"
+	"gopkg.in/bblfsh/sdk.v2/uast/nodes"
+)
+
+type decisionType interface {
+	isDecisionType()
+	cost() int
+}
+
+type basicDecision struct {
+	privateCost int
+}
+
+func (basicDecision) isDecisionType() {}
+
+func (b basicDecision) cost() int { return b.privateCost }
+
+// match decision types together with their params
+type sameDecision struct{ basicDecision }
+type replaceDecision struct{ basicDecision }
+type matchDecision struct{ basicDecision }
+type permuteDecision struct {
+	basicDecision
+	permutation []int
+}
+
+//convinience method for choosing the cheapest decision
+func min(self, candidate decisionType) decisionType {
+	if self.cost() > candidate.cost() {
+		return candidate
+	} else {
+		return self
+	}
+}
+
+//type for cache
+type keyType struct{ k1, k2 ID }
+
+//cache for diff computation
+type cacheStorage struct {
+	decisions map[keyType]decisionType
+	sizes     map[ID]int
+	changes   Changelist
+}
+
+func emptyCacheStorage() cacheStorage {
+	return cacheStorage{
+		make(map[keyType]decisionType),
+		make(map[ID]int),
+		nil,
+	}
+}
+
+// caches (for perf) tree size (node count)
+func (ds *cacheStorage) nodeSize(n nodes.Node) int {
+	label := nodes.UniqueKey(n)
+	if cnt, ok := ds.sizes[label]; ok {
+		return cnt
+	}
+	ret := nodes.Count(n, nodes.KindsNotNil)
+	ds.sizes[label] = ret
+	return ret
+}
+
+// find the cheapest way to naively match src and dst and return the action with its combined cost
+func (ds *cacheStorage) decideAction(src, dst nodes.Node) decisionType {
+	label := keyType{nodes.UniqueKey(src), nodes.UniqueKey(dst)}
+
+	if val, ok := ds.decisions[label]; ok {
+		return val
+	}
+
+	// one can always just create the dst ignoring the src
+	cost := ds.nodeSize(dst) + 1
+
+	var bestDecision decisionType
+	bestDecision = replaceDecision{basicDecision{cost}}
+
+	if nodes.KindOf(src) != nodes.KindOf(dst) {
+		ds.decisions[label] = bestDecision
+		return bestDecision
+	}
+
+	switch src.(type) {
+
+	case nodes.Value:
+		src, dst := src.(nodes.Value), dst.(nodes.Value)
+		cost = 0
+		if src == dst {
+			bestDecision = min(bestDecision, sameDecision{basicDecision{cost}})
+		} else {
+			cost = 1
+			bestDecision = min(bestDecision, replaceDecision{basicDecision{cost}})
+		}
+
+	case nodes.Object:
+		src, dst := src.(nodes.Object), dst.(nodes.Object)
+		cost = 0
+
+		keys := make(map[string]bool)
+		iterate := func(keyset nodes.Object) {
+			for key := range keyset {
+				if in := keys[key]; !in {
+					keys[key] = true
+					cost += ds.decideAction(src[key], dst[key]).cost()
+				}
+			}
+		}
+		iterate(src)
+		iterate(dst)
```

Comment:
You are not resetting `keys` here. Is it intentional?

Also, you can call `Keys()` on both objects and scan these slices (they are sorted by key).

Reply:
not resetting keys is the point. I want to visit every key from the joint key set exactly once. Haven't found a better method in Go.

Reply:
What is Keys() function? Can't find any docs for it.

Also, added better comments.

Reply:
`Keys` is a function on `nodes.Object`